### ISSUE TITLE
[Feature Request] QoL Updates

### ISSUE THREAD

**Issue by thelqh** [ShaneIsrael/fireshare#264]
Hi Shane, love the project and love using it all the time.

Would love to have these features considered:

- Video quality selector
- Video metadata and info (file size, bitrate, etc.) backend display
- Multi-user access (currently using multiple docker instances)
- Password protection for private videos

Thanks!

**Comment by Krojack** [comment]
HEVC (x265) playback support would be nice but not sure if possible with licensing or what not. AV1 playback does work but takes long to encode via CPU if you don't have a GPU that supports it.

**Comment by thelqh** [comment]
> HEVC (x265) playback support would be nice but not sure if possible with licensing or what not. AV1 playback does work but takes long to encode via CPU if you don't have a GPU that supports it.

I was moreso thinking 480p, 720p, etc. so those on slower internet connections can stream it more. I'm using it with Cloudflare Tunnels so it might also improve compatability there for those who are getting blocked.

**Comment by J-Stuff** [comment]
I'd be keen to rework some of the backend to support HLS video streaming and transcoded videos on upload. However knowing how the backend works right now it would be a breaking change requiring either a migration setup or a complete instance reset upon upgrade.

**Comment by J-Stuff** [comment]
At the same time I'd likely fix #267

**Comment by J-Stuff** [comment]
Wow, been a crazy few months here for me.

Here's what I was thinking of:

Changing how the backend stores video, by moving everything into subfolders for the project to work with:
```
data/
├─ video/
│  ├─ UNIQUE_V_ID_1/
│  │  ├─ transcoded/
│  │  │  ├─ 1080p/
│  │  │  │  ├─ part.1
│  │  │  │  ├─ part.2
│  │  │  │  ├─ part.3
│  │  │  ├─ 720p/
│  │  │  ├─ ETC ETC.../
│  │  ├─ original.mp4
│  │  ├─ metadata.json
│  ├─ UNIQUE_V_ID_2/
│  ├─ UNIQUE_V_ID_3/
```

On the python backend side, would need to change how and where the metadata is generated and stored. Additionally, probably also looking at adding an additional directory to the docker container (something like `/upload`) to allow for docker binds to retrieve new video uploads from a directory.

A change to how the /_content/ path is served from the NGINX backend would also need to be made. As well as several frontend changes to ad additional settings (Like how many different transcoded versions of each video should be made (trade off for storage usage and how many different quality settings you'd have to chose from)) the video player would have to be changed to be HLS.js compliant as it's not standard for browsers to have this compatibility baked in. As well as changes made to the Open Graph metadata tags for previews on Discord, Facebook ETC.

Lastly you'd also need to add a script in the "bootstrap" code for fireshare preventing it from booting if it detected a legacy DB or media layout to prevent bricking everyone's instances (Because as bad a practice as it is, a lot of people deploy using the :latest image tag and might not realise such a breaking update has happened.)

@ShaneIsrael would love to get some feedback/input on this, as it is such a huge breaking change and it is your repo ^^

**Comment by ShaneIsrael** [comment]
@J-Stuff Sorry for the late reply (been in the middle of moving into a new house) this is all very reasonable in order to get a transcoding option in place. 

Adding transcoding is a rather large undertaking and since it was never a consideration when we first started working on this we didn't design much of it with that in mind. 

Now in terms of handling the whole latest image issue and breaking changes... A script preventing booting would probably need to be the bare minimum but then that introduces other issues, mainly being that most people are not thoughtful enough to look at logs when something isnt working and may just assume that there stuff is broken and possibly manually delete there database in an attempt to reset/fix it.

We could update fireshare to fetch a "banner" message that is hosted on github, and then we could have it say 
"Fireshare is being updated, the latest image tag will soon cause a breaking change. If you do not want to update, make sure you pin to version x.x.x" 

Then after a month or two we can stop updating the latest image to give people time. 

Once these breaking changes are in place though, I do think it would be a good idea to have a script to check for and archive the old database in case somebody wanted to revert back rather than destroying it.

**Comment by ShaneIsrael** [comment]
@J-Stuff Yeah that is fine but I wouldn't read it directly from the codebase but rather the raw github url to that file. That way changes made to that file do not require a new container release and pull.

**Comment by theotherguy2175** [comment]
@badgermarten I could lookinto the meta data feature. What all would you be looking for? Might be pretty easy to implement. 

Would you want a button on each video when clicked to show the metadata? Or like a admin settings toggle to show it for every video?
